### PR TITLE
Decomposition of bmm and mm for dot product

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary: Decomposition of bmm and mm for dot product

Test Plan: sandcastle, and Bert

Differential Revision: D48055141

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Review by jansel — APPROVED** [review]

**Comment by hl475** [comment]
fix the lint https://github.com/pytorch/pytorch/actions/runs/5762639592/job/15623143501?pr=106593 ?

**Comment by mikekgfb** [comment]
D47969739<https://www.internalfb.com/diff/D47969739>.

Also had some questions on Cuda test fails where we expect a specific number of torch.mm instances when executing on gpu D47881983<https://www.internalfb.com/diff/D47881983> which correspond to #106219

Should we fix up expected counts, or limit dot product and other expansions to cpu only?



Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by jansel** [comment]
> Should we fix up expected counts, or limit dot product and other expansions to cpu only?

Updating the expected counts is fine, but we should check perf on CUDA.  I believe this will result in us not using tensor cores (since we won't generate tl.dot) -- but that might not matter since these may be memory bound.

**Comment by mikekgfb** [comment]
> > Should we fix up expected counts, or limit dot product and other expansions to cpu only?
> 
> Updating the expected counts is fine, but we should check perf on CUDA. I believe this will result in us not using tensor cores (since we won't generate tl.dot) -- but that might not matter since these may be memory bound.

I'm holding off on committing simplifications for CUDA to better understand the code generation choices, and focus on CPU computations where for small vector sizes, the call to the subroutine is longer than the computation itself.

For non-CPU devices we'll want to understand whether it's better breaking up operators is better, or leaving them as tensor ops and special casing e.g. dot-prod during code gen.  Potentially also setting the boundary when one or the other strategy is chosen based on platform characteristics because we can't tangibly incorporate all rules for all targets.  (Plus, some like CPU or CUDA have different implementations across vendors and even incompatible architectures, e.g., x86 vs ARM and Nvidia vs. AMD) 

Totally plausible to do this as a config file that has a view optimization policies, see also a parallel discussion with @jgong5 on how to set the decisions for expansions.

File: test/test_mmdecomp.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    """
+    previous_mode: bool = torch.are_deterministic_algorithms_enabled()
+    previous_warn_only: bool = torch.is_deterministic_algorithms_warn_only_enabled()
+    try:
+        torch.use_deterministic_algorithms(mode, warn_only = warn_only)
+        yield {}
+    finally:
+        torch.use_deterministic_algorithms(previous_mode, warn_only = previous_warn_only)
+
+
```
Comment (malfet) [inline test/test_mmdecomp.py:52]: Isn't it the same as `torch.testing._internal.DeterministicGuard`? Besides, it's unused, so why not delete it?
```suggestion
```

Comment (mikekgfb) [reply test/test_mmdecomp.py:52]: Copy pasta from test_transformer.py ;)

**Comment by mikekgfb** [comment]
What's the best path forward with bfloat16 tests?  

At present, no matter how high we make atol/rtol (bumped them up to 1e+3 which begs the question whether these tests are even still useful), there are always some outliers that are not "sufficiently close".  Example => https://github.com/pytorch/pytorch/actions/runs/5824930571/job/15795615962

which fails with the following violation:
```
FAILED [0.0311s] test_mmdecomp.py::TestDecompCPU::test_batched_mm_bfloat16_bs_10_cpu_bfloat16 - AssertionError: Tensor-likes are not close!

Mismatched elements: 1408 / 10240 (13.8%)
Greatest absolute difference: 2.6185791516962843e+38 at index (4, 30, 26) (up to 3000.0 allowed)
Greatest relative difference: inf at index (4, 30, 26) (up to 480.0 allowed)
```

How should we think about this data type for end users and expected result testing as a function of transformations a compiler might make?

cc: @jgong5 @jansel @ezyang @malfet

**Comment by mikekgfb** [comment]
@pytorchbot merge


Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Pull request closed without merging**